### ISSUE TITLE
OOM for big loops(>170k)

### ISSUE THREAD

**Issue by doraeimo** [mockito/mockito#1825]
Recently I'm upgrading Mockito from 1.10 to 2.25.

There's a failure test case (which was good in version 1.10) that caused by OutOfMemory(Xmx1024m) when it runs a big loop(times > 170k).

I've checked the dump file and found too many StackTraceElement instances there.
Also noticed that for any mock object, it has a property chain of

> mockObj.mockitoInterceptor.handler.mockHandler.delegate.invocationContainer.registeredInvocations.invocations  // a LinkedList which will increase after each loop

And there's a StackTraceElement here
> invocations[n].location.stackTraceHolder

I know there's a similar property chain in Mockito 1.10.
Not sure why it goes OOM after the upgrade(looks like a memory leak).
Please verify this. Thanks!

**Comment by bric3** [comment]
This is not a leak, Mockito stores this information on purpose to identify the location where the stub has been called for verification and of strict mode. In Mockito 2 offers more precise diagnostics thanks to this data.

However the API acknowledges that some use case just need to be stubbed, and allow to skip this data. [`withSettings().stubOnly()`](https://www.javadoc.io/static/org.mockito/mockito-core/2.28.2/org/mockito/MockSettings.html#stubOnly--)

**Comment by doraeimo** [comment]
I tried jvisualvm, observed that more loops it did,
- for Mockito 1.10, it costed 150m memory on average, and went stable after each GC.(Also stored new StackTraceElement instance after each loop)
- for Mockito 2.25, it went out of memory soon.

If it is not a leak, I hope there's a clear hint or warning to tell us where is the suspicious test case, and sth like "avoid big loops". OOM is not very friendly. Thanks.

**Comment by bric3** [comment]
Indeed  OOMs are not very friendly.
Do you have an idea how to do that efficiently and for every configuration?

-- Brice

**Comment by TimvdLippe** [comment]
Am I to understand correctly you are creating 170k mocks in a for-loop?

**Comment by doraeimo** [comment]
> Am I to understand correctly you are creating 170k mocks in a for-loop?

No, it goes like this,

```
when(mockObj.foo()).thenReturn(bar);
for(;;){ //loop times > 170k
    mockObj.foo()
}
```

**Comment by bric3** [comment]
@doraeimo did you tried to set `stubOnly()` ?

**Comment by TimvdLippe** [comment]
I am fairly certain this was resolved with #1683 and verified with a test in https://github.com/mockito/mockito/tree/main/subprojects/memory-test
